### PR TITLE
pkgsCross.musl64.go: fix build

### PR DESCRIPTION
**Pull request by yu-re-ka**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by zowoq — CHANGES_REQUESTED** [review]
Same change needs to be applied to 1.18 as well.

Needs to go to staging due to rebuilds.

**Comment by yu-re-ka** [comment]
Weird, why does it cause rebuilds on darwin, but not on linux?

**Comment by zowoq** [comment]
> Weird, why does it cause rebuilds on darwin, but not on linux?

Yeah, we should find out why before we merge.

**Comment by superherointj** [comment]
Using this patch at `linux-x86_64`, `pkgsCross.musl64.go` builds (as correctly stated). But `pkgsMusl.go` does not. Is this expected? Any idea why it happens?

**Comment by yu-re-ka** [comment]
pkgsMusl.go is out of scope for this PR

**Comment by superherointj** [comment]
@yu-re-ka Anything else stopping this PR?

If it is just rebuilding for darwin and is not breaking anything, I guess targeting staging would be fine.
Darwin should not be stopping us from having a working go musl compiler... I cannot test it. =\

Can you replicate changes to 1.18 as zowoq requested?

**Comment by yu-re-ka** [comment]
No, we can not merge this without knowing why it creates all those darwin rebuilds

File: pkgs/development/compilers/go/1.19.nix
Comment (sternenseemann) [inline pkgs/development/compilers/go/1.19.nix:144]: If the intention is to allow this code here as it runs stuff built for host during the built, then this should be made clear properly (and you gain some cases not covered for free):

```suggestion
  '' + (if !(stdenv.buildPlatform.canExecute stdenv.hostPlatform) then ''
```

File: pkgs/development/compilers/go/1.19.nix
Comment (sternenseemann) [inline pkgs/development/compilers/go/1.19.nix:150]:
```suggestion
  '' else if !(stdenv.buildPlatform.canExecute stdenv.targetPlatform) then ''
```

**Review by sternenseemann — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/go/1.19.nix
Comment (yu-re-ka) [reply pkgs/development/compilers/go/1.19.nix:144]: No, the problem is that when the go bootstrap tooling _thinks_ it's cross-compiling, it will put the result into a subdirectory /$target/.

**Comment by superherointj** [comment]
@Et7f3 Unrelated. But as you just shown your debugging skills:
I remember a Darwin mass rebuild when trying to fix the go compiler for pkgsCross.
Maybe you have a hint?

(It's a fairly simple change.)

**Comment by yu-re-ka** [comment]
Rebased, still 1000+ rebuilds on darwin

File: pkgs/development/compilers/go/1.20.nix
Comment (zowoq) [inline pkgs/development/compilers/go/1.20.nix:151]: Before this change `rm -rf bin/*_*` was included in the `preInstall` only for `x86_64-darwin`.

No idea why it was like that but it wasn't correct and it was a no op anyway as there wasn't anything for it to remove.

I think it's fine to merge this but I'd send to to staging so we aren't causing a pointless rebuild of all `x86_64-darwin` go packages on master.

Comment (yu-re-ka) [reply pkgs/development/compilers/go/1.20.nix:151]: Thanks for figuring out what the difference is and confirming that it's fine to merge

**Pull request merged**
